File: common/protocol.py
"""
应用层通信协议定义。

设计目标(详见 docs/architecture.md):
- 握手阶段(密钥交换 kex_init / kex_reply)使用明文 JSON 文本帧,
  因为此时双方尚未协商出会话密钥。
- 握手完成后,所有消息(无论控制信令还是视频帧)一律封装为二进制帧,
  并使用 AES-256-GCM 加密(见 common/crypto.py),中转服务器 relay 只
  转发不可读的密文字节,天然具备端到端加密属性。

二进制帧(加密前的明文)结构:
    [1 字节 msg_kind][payload]

msg_kind:
    MSG_CONTROL (0x01)     -> payload 是 UTF-8 JSON 字节串,字段 "t" 表示消息类型
    MSG_VIDEO_FRAME (0x02) -> payload 是 FRAME_HEADER_STRUCT 头部 + 图像二进制数据

视频帧头部(FRAME_HEADER_STRUCT, 16 字节, 大端):
    magic(1)=0xF1, seq(uint32), ts_ms(uint32, 会话相对时间戳),
    width(uint16), height(uint16), quality(uint8, 1-100),
    fmt(uint8, 0=JPEG 1=WEBP), flags(uint8, bit0=关键帧/完整帧)
"""
from __future__ import annotations

import json
import struct
from dataclasses import dataclass
from typing import Any

PROTOCOL_VERSION = 1

# ---- 明文握手阶段消息类型(仅用于 kex 之前,走 WebSocket 文本帧)----
# 握手只有 host -> client 单向的 kex_init(内含 salt + KDF 参数);client 收到后
# 直接本地派生密钥,无需再回复握手消息,详见 common/crypto.py 顶部说明。
KEX_INIT = "kex_init"
KEX_REJECT = "kex_reject"  # relay/host 侧在无法配对或版本不兼容/繁忙时明文告知原因

# ---- 加密后二进制帧的一级分类 ----
MSG_CONTROL = 0x01
MSG_VIDEO_FRAME = 0x02

# ---- MSG_CONTROL JSON 内的 "t" 字段取值 ----
T_HELLO = "hello"
T_HELLO_ACK = "hello_ack"
T_PING = "ping"
T_PONG = "pong"
T_FRAME_ACK = "frame_ack"
T_MOUSE_MOVE = "mmove"
T_MOUSE_BUTTON = "mbtn"
T_MOUSE_SCROLL = "mscroll"
T_KEY = "key"
T_CLIPBOARD = "clip"
T_QUALITY_SET = "quality"
T_STATS = "stats"
T_ERROR = "error"
T_BYE = "bye"

# ---- 画质档位(智能模式在这些档位之间自动切换;手动模式由用户直接选定)----
QUALITY_AUTO = "auto"
QUALITY_CUSTOM = "custom"
QUALITY_PRESETS = ("extreme", "smooth", "balanced", "clear", "hd")  # 由低到高

FRAME_HEADER_STRUCT = struct.Struct(">BIIHHBBB")  # magic,seq,ts_ms,w,h,quality,fmt,flags
FRAME_MAGIC = 0xF1
FMT_JPEG = 0
FMT_WEBP = 1
FLAG_KEYFRAME = 0b0000_0001

MAX_CONTROL_MESSAGE_BYTES = 64 * 1024  # 控制消息(JSON)上限,防止异常输入拖垮解析
MAX_CLIPBOARD_CHARS = 64 * 1024  # 单次剪贴板同步的最大字符数


class ProtocolError(ValueError):
    """协议格式错误(收到无法解析/不合法的帧)。"""


def encode_control(msg: dict[str, Any]) -> bytes:
    """将控制消息字典编码为待加密的明文 payload(带 msg_kind 前缀)。"""
    body = json.dumps(msg, ensure_ascii=False, separators=(",", ":")).encode("utf-8")
    if len(body) > MAX_CONTROL_MESSAGE_BYTES:
        raise ProtocolError(f"control message too large: {len(body)} bytes")
    return bytes([MSG_CONTROL]) + body


def encode_video_frame(
    *, seq: int, ts_ms: int, width: int, height: int, quality: int,
    fmt: int, keyframe: bool, image_bytes: bytes,
) -> bytes:
    """将一帧视频数据编码为待加密的明文 payload(带 msg_kind 前缀)。"""
    flags = FLAG_KEYFRAME if keyframe else 0
    header = FRAME_HEADER_STRUCT.pack(
        FRAME_MAGIC, seq & 0xFFFFFFFF, ts_ms & 0xFFFFFFFF,
        width, height, quality, fmt, flags,
    )
    return bytes([MSG_VIDEO_FRAME]) + header + image_bytes


@dataclass
class VideoFrame:
    seq: int
    ts_ms: int
    width: int
    height: int
    quality: int
    fmt: int
    keyframe: bool
    image_bytes: bytes


def decode_plaintext(plaintext: bytes) -> tuple[int, dict[str, Any] | VideoFrame]:
    """解码解密后的明文帧,返回 (msg_kind, 内容)。"""
    if len(plaintext) < 1:
        raise ProtocolError("empty plaintext frame")
    kind = plaintext[0]
    body = plaintext[1:]
    if kind == MSG_CONTROL:
        if len(body) > MAX_CONTROL_MESSAGE_BYTES:
            raise ProtocolError("control message too large")
        try:
            msg = json.loads(body.decode("utf-8"))
        except (UnicodeDecodeError, json.JSONDecodeError) as exc:
            raise ProtocolError(f"invalid control json: {exc}") from exc
        if not isinstance(msg, dict) or "t" not in msg:
            raise ProtocolError("control message missing 't' field")
        return kind, msg
    if kind == MSG_VIDEO_FRAME:
        if len(body) < FRAME_HEADER_STRUCT.size:
            raise ProtocolError("video frame header truncated")
        magic, seq, ts_ms, width, height, quality, fmt, flags = FRAME_HEADER_STRUCT.unpack(
            body[: FRAME_HEADER_STRUCT.size]
        )
        if magic != FRAME_MAGIC:
            raise ProtocolError("bad video frame magic")
        image_bytes = body[FRAME_HEADER_STRUCT.size :]
        return kind, VideoFrame(
            seq=seq, ts_ms=ts_ms, width=width, height=height,
            quality=quality, fmt=fmt, keyframe=bool(flags & FLAG_KEYFRAME),
            image_bytes=image_bytes,
        )
    raise ProtocolError(f"unknown msg_kind: {kind}")


def encode_kex_init(salt_b64: str, iterations: int) -> str:
    return json.dumps({
        "t": KEX_INIT, "v": PROTOCOL_VERSION,
        "salt": salt_b64, "kdf": "pbkdf2-sha256", "iterations": iterations,
    }, ensure_ascii=False)


def parse_kex_text(text: str) -> dict[str, Any]:
    if len(text) > 8192:
        raise ProtocolError("kex message too large")
    try:
        msg = json.loads(text)
    except json.JSONDecodeError as exc:
        raise ProtocolError(f"invalid kex json: {exc}") from exc
    if not isinstance(msg, dict) or "t" not in msg:
        raise ProtocolError("kex message missing 't' field")
    return msg

File: client/static/js/protocol.js
'use strict';
/*
 * 协议常量与二进制帧解析。
 *
 * 与 Python 侧 common/protocol.py 逐字节对应,任何一边改动都必须同步另一边。
 *
 * 加密后的每条 WebSocket 二进制消息 = [12字节随机nonce][AES-GCM密文]。
 * 解密后的明文第 1 个字节是 msg_kind:
 *   1 = 控制消息(UTF-8 JSON)
 *   2 = 关键帧    -> [16字节帧头][整幅JPEG]
 *   3 = 增量帧    -> [16字节帧头][矩形数(2)][若干 [12字节矩形头][JPEG] ]
 *   4 = 文件分块  -> [transfer_id(4)][seq(4)][原始字节]
 *
 * 帧头(大端): magic(1)=0xF1 seq(4) ts_ms(4) width(2) height(2)
 *              quality(1) fmt(1) flags(1, bit0=关键帧)
 * 矩形头(大端): x(2) y(2) w(2) h(2) size(4)
 * 矩形坐标位于"输出图像坐标系",即与最近一个关键帧声明的 width/height 同一
 * 坐标系,客户端直接把每个矩形画到 canvas 对应位置即可,无需坐标换算。
 */

export const MSG_CONTROL = 1;
export const MSG_VIDEO_FRAME = 2;
export const MSG_VIDEO_DELTA = 3;
export const MSG_FILE_CHUNK = 4;

export const FRAME_MAGIC = 0xf1;
export const FRAME_HEADER_BYTES = 16;
export const RECT_HEADER_BYTES = 12;
export const FILE_CHUNK_HEADER_BYTES = 8;

export const FMT_JPEG = 0;
export const FMT_WEBP = 1;
export const MIME_BY_FMT = { 0: 'image/jpeg', 1: 'image/webp' };

export const MAX_FRAME_DIMENSION = 8192; // 防止被篡改的宽高把 canvas/内存撑爆
export const MAX_RECTS_PER_FRAME = 512;
export const FILE_CHUNK_BYTES = 64 * 1024;

export class ProtocolError extends Error {}

function readFrameHeader(dv) {
  if (dv.getUint8(0) !== FRAME_MAGIC) throw new ProtocolError('帧头 magic 不匹配');
  const header = {
    seq: dv.getUint32(1),
    tsMs: dv.getUint32(5),
    width: dv.getUint16(9),
    height: dv.getUint16(11),
    quality: dv.getUint8(13),
    fmt: dv.getUint8(14),
    keyframe: (dv.getUint8(15) & 0x01) !== 0,
  };
  if (header.width === 0 || header.height === 0
      || header.width > MAX_FRAME_DIMENSION || header.height > MAX_FRAME_DIMENSION) {
    throw new ProtocolError(`帧尺寸异常: ${header.width}x${header.height}`);
  }
  return header;
}

/** 解析关键帧。body 是去掉 msg_kind 之后的字节。 */
export function decodeKeyframe(body) {
  if (body.length < FRAME_HEADER_BYTES) throw new ProtocolError('关键帧帧头被截断');
  const dv = new DataView(body.buffer, body.byteOffset, body.byteLength);
  const header = readFrameHeader(dv);
  return { ...header, imageBytes: body.subarray(FRAME_HEADER_BYTES) };
}

/** 解析增量帧,返回 { ...帧头, rects: [{x,y,w,h,imageBytes}] }。 */
export function decodeDelta(body) {
  if (body.length < FRAME_HEADER_BYTES + 2) throw new ProtocolError('增量帧帧头被截断');
  const dv = new DataView(body.buffer, body.byteOffset, body.byteLength);
  const header = readFrameHeader(dv);

  const rectCount = dv.getUint16(FRAME_HEADER_BYTES);
  if (rectCount > MAX_RECTS_PER_FRAME) throw new ProtocolError(`增量帧矩形数过多: ${rectCount}`);

  const rects = [];
  let cursor = FRAME_HEADER_BYTES + 2;
  for (let i = 0; i < rectCount; i++) {
    if (body.length < cursor + RECT_HEADER_BYTES) throw new ProtocolError('增量帧矩形头被截断');
    const x = dv.getUint16(cursor);
    const y = dv.getUint16(cursor + 2);
    const w = dv.getUint16(cursor + 4);
    const h = dv.getUint16(cursor + 6);
    const size = dv.getUint32(cursor + 8);
    cursor += RECT_HEADER_BYTES;
    if (body.length < cursor + size) throw new ProtocolError('增量帧矩形数据被截断');
    rects.push({ x, y, w, h, imageBytes: body.subarray(cursor, cursor + size) });
    cursor += size;
  }
  return { ...header, rects };
}

/** 把一个文件分块编码成待加密的明文(含 msg_kind 前缀)。 */
export function encodeFileChunk(transferId, seq, data) {
  const out = new Uint8Array(1 + FILE_CHUNK_HEADER_BYTES + data.length);
  const dv = new DataView(out.buffer);
  out[0] = MSG_FILE_CHUNK;
  dv.setUint32(1, transferId >>> 0);
  dv.setUint32(5, seq >>> 0);
  out.set(data, 1 + FILE_CHUNK_HEADER_BYTES);
  return out;
}

/** 解析文件分块(被控端 -> 主控端方向的下载)。 */
export function decodeFileChunk(body) {
  if (body.length < FILE_CHUNK_HEADER_BYTES) throw new ProtocolError('文件分块头部被截断');
  const dv = new DataView(body.buffer, body.byteOffset, body.byteLength);
  return {
    transferId: dv.getUint32(0),
    seq: dv.getUint32(4),
    data: body.subarray(FILE_CHUNK_HEADER_BYTES),
  };
}

/**
 * 探测浏览器能解码哪些格式,用于在 hello 里向被控端声明能力。
 * WebP 在增量帧的小矩形上比 JPEG 省 80% 以上,但 Safari 14 以下不支持,
 * 必须能优雅回落到 JPEG。
 */
export async function detectCodecs() {
  const codecs = ['jpeg'];
  try {
    // 1x1 的合法 WebP;能成功解码才说明真的支持
    const bytes = Uint8Array.from(atob(
      'UklGRhoAAABXRUJQVlA4TA0AAAAvAAAAEAcQERGIiP4HAA=='), (c) => c.charCodeAt(0));
    const bitmap = await createImageBitmap(new Blob([bytes], { type: 'image/webp' }));
    bitmap.close();
    codecs.unshift('webp');
  } catch {
    // 不支持 WebP,保持只声明 jpeg
  }
  return codecs;
}
